### PR TITLE
feat(validator): Add type validator constraint

### LINKED ISSUE #169: [validator] Add built-in `type` basic validator constraint
We should add a built-in `type` constraint under "basic" constraints to cover runtime type checks in the validator package.

## Proposal

Add a new built-in constraint named `type` with options shaped like the existing built-ins, including shared validator metadata such as `groups`:

```ts
type TypeOptions = ConstraintOptions & {
  type: string | string[];
  message?: string;
};
```

Example usage:

```ts
const rules = {
  age: [{ type: { type: 'number' } }],
  tags: [{ type: { type: 'array' } }],
  name: [{ type: { type: 'string', groups: ['create'] } }],
  identifier: [{ type: { type: ['string', 'number'] } }],
};
```

## Expected behavior

- `type` should live in the `basic` constraint group.
- It should be exported from the validator constraints public API and registered in `builtInConstraints`.
- It should accept a required `type` option as either a single allowed type or a list of allowed types.
- It should support shared validator metadata exposed through `ConstraintOptions`, including `groups`.
- It should return no violations when the value matches at least one allowed type.
- It should return one violation when the value matches none of the allowed types.
- It should support a custom `message` option, consistent with the existing constraints.
- It should preserve current optional-field semantics used by other constraints: `undefined` should be considered valid unless presence is enforced separately.
- `null` handling should be defined explicitly and covered by tests.

## Suggested initial supported type aliases

- `string`
- `number`
- `boolean`
- `bigint`
- `symbol`
- `function`
- `object`
- `array`
- `null`
- `undefined`

## Default message

A sensible default would be:

```txt
This value should be of type <expected type>.
```

For multiple allowed types, the message can mention the full allowed set.

### PR DESCRIPTION
**Pull request by EmanFateen**
### Overview

In this PR, I have introduced new validation constraint called type. 
```ts
type TypeOptions = ConstraintOptions & {
  type: AllowedTypes | AllowedTypes[];
  message?: string;
};
```
```ts
export type AllowedTypes =
  | 'string'
  | 'number'
  | 'boolean'
  | 'bigint'
  | 'symbol'
  | 'function'
  | 'object'
  | 'array'
  | 'null';
```

It's exposed through the `builtInConstraints`   and can be used like 
```ts
const rules = {
  age: [{ type: { type: 'number' } }],
  identifier: [{ type: { type: ['string', 'number'] } }],
}
```

- Main purpose of this constraint is to validate whether the given value matches the expected type or one of the expected types.
- This constraint is concerned about the type of the given value and should work as a required constraint, which means  is the value is undefined the constraint will return no violations.
- Last point doesn't applied for the `null` as `null` is a real value. 

### Changes made: 
1. Added the `type` constraint under basic constraint. 
2. Covered the required cases in `type.test` file. 
3. Exposed the `type` constraint through the `builtInConstraints` object.

| Q       | A                      |
| ------- | ---------------------- |
| License | GPLv3                  |
| Issue   | Closes #169 > |

### CODE REVIEW

**Review by imdhemy — CHANGES_REQUESTED** [review]
~~This PR introduces regression to the validator component’s structure. We’ve already improved the structure in #175. Could you kindly modify the changes here to align with the updated structure? @EmanFateen~~

I was reviewing an older version of the PR.

File: src/validator/constraints/basic/type.test.ts [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+      runNestedRules: () => [],
+    };
+
+    const violations = type(false, context);
+
+    expect(violations).toHaveLength(1);
+    expect(violations[0]).toEqual({
+      path: 'identifier',
+      constraint: 'type',
+      message: 'This value should be of type string or number.',
```
Comment (imdhemy) [inline src/validator/constraints/basic/type.test.ts]: This format has two minor issues:

1. How would you format a message for more than two types?

For example, `This value should be of type string, number, or boolean.` Have you already addressed this grammar issue? If yes, we need to include a test case to cover it.

2. To handle the grammar corner case, you’ll need to add more code.

I prefer a simpler format that requires less code and reduces maintenance effort.

For instance, `This value should match at least one of these types [string, number, boolean].`

File: src/validator/constraints/basic/type.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
+import { ConstraintOptions } from '@/validator/constraint';
+import { ConstraintContext } from '@/validator/constraint-validator';
+import { Violation } from '@/validator/violation';
+
+const DEFAULT_MESSAGE = 'This value should be of type {type}.';
```
Comment (imdhemy) [inline src/validator/constraints/basic/type.ts]: We need two default messages in this constraint: one for singular and one for plural, to cover the two cases of using a single or multiple type.

File: src/validator/constraints/basic/type.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+export type TypeOptions = ConstraintOptions & {
+  type: string | string[];
+  message?: string;
+};
+
+export function type(value: unknown, context: ConstraintContext<TypeOptions>): Violation[] {
+  if (value === undefined) {
```
Comment (imdhemy) [inline src/validator/constraints/basic/type.ts]: This case is already covered in the test cases, but it’s better to have a separate test case for it, similar to how we did for the `email` constraint. This approach clearly communicates the intended behavior.

File: src/validator/constraints/basic/type.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    return [];
+  }
+
+  const options = context.options;
+  const expectedTypes = Array.isArray(options.type) ? options.type : [options.type];
+
+  if (matchesExpectedType(value, expectedTypes)) return [];
```
Comment (imdhemy) [inline src/validator/constraints/basic/type.ts]: Argument type `string | [string | string[]]` is not assignable to parameter type `string[]`

File: src/validator/constraints/basic/type.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+  return [
+    {
+      path: context.path,
+      message: options.message ?? getDefaultMessageWith(expectedTypes),
```
Comment (imdhemy) [inline src/validator/constraints/basic/type.ts]: Argument type `string | [string | string[]]` is not assignable to parameter type `string[]`

File: src/validator/constraints/basic/type.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
+import { ConstraintOptions } from '@/validator/constraint';
+import { ConstraintContext } from '@/validator/constraint-validator';
+import { Violation } from '@/validator/violation';
+
+const DEFAULT_MESSAGE = 'This value should be of type {type}.';
+
+export type TypeOptions = ConstraintOptions & {
+  type: string | string[];
```
Comment (imdhemy) [inline src/validator/constraints/basic/type.ts]: We could provide a specific list of the types we can verify instead of using a generic string or a string list.

**Review by imdhemy — CHANGES_REQUESTED** [review]
@EmanFateen 
Thanks for moving this forward. The implementation is good and covers the issue, we just need to tighten the types and simplify the implementation.

File: src/validator/constraints/basic/type.test.ts (hunk 1 above)
Comment (EmanFateen) [reply src/validator/constraints/basic/type.test.ts]: I handled it with this format `This value should match at least one of these types [string, number, boolean].`

File: src/validator/constraints/basic/type.ts (hunk 2 above)
Comment (EmanFateen) [reply src/validator/constraints/basic/type.ts]: Done

File: src/validator/constraints/basic/type.ts (hunk 3 above)
Comment (EmanFateen) [reply src/validator/constraints/basic/type.ts]: Done.

File: src/validator/constraints/basic/type.ts (hunk 5 above)
Comment (EmanFateen) [reply src/validator/constraints/basic/type.ts]: I didn't get this commit.  Would you please explain more?

File: src/validator/constraints/basic/type.ts (hunk 6 above)
Comment (EmanFateen) [reply src/validator/constraints/basic/type.ts]: Done

Comment (imdhemy) [reply src/validator/constraints/basic/type.ts]: The allowed types should be strings.
```diff
- string | number | boolean ...
+ 'string' | 'number' | 'boolean' ...
```

File: src/validator/constraints/basic/type.ts (hunk 5 above)
Comment (imdhemy) [reply src/validator/constraints/basic/type.ts]: It’s an issue with the declared types. You’ve already changed the code, but unfortunately, we have other issues with the types.

File: src/validator/constraints/basic/type.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+  type: AllowedTypes | AllowedTypes[];
+  message?: string;
+};
+
+export function type(value: unknown, context: ConstraintContext<TypeOptions>): Violation[] {
+  if (value === undefined) {
+    return [];
+  }
+
+  const options: TypeOptions = context.options;
```
Comment (imdhemy) [inline src/validator/constraints/basic/type.ts]: I don’t understand the necessity of this intermediate step, but I suppose things will become clearer once you enhance the types.

File: src/validator/constraints/basic/type.ts [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+function getValueType(value: unknown): string {
+  if (value === null) return 'null';
+  if (Array.isArray(value)) return 'array';
+  return typeof value;
+}
+
+function getDefaultMessageWith(types: AllowedTypes[]): string {
+  if (Array.isArray(types) && types.length > 1) {
```
Comment (imdhemy) [inline src/validator/constraints/basic/type.ts]: Since you’ve already declared `types` as an array, there’s no need to perform the check.

File: src/validator/constraints/basic/type.ts (hunk 6 above)
Comment (imdhemy) [reply src/validator/constraints/basic/type.ts]: Also undefined is not an allowed type. `undefined` should be bypassed.

**Review by imdhemy — CHANGES_REQUESTED** [review]

File: src/validator/constraints/basic/type.ts [hunk 9]
```diff
@@ -0,0 +1,50 @@
+import { ConstraintOptions } from '@/validator/constraint';
+import { ConstraintContext } from '@/validator/constraint-validator';
+import { Violation } from '@/validator/violation';
+
+const DEFAULT_MESSAGE = 'This value should be of type {type}.';
+const DEFAULT_MESSAGE_FOR_MULTIPLE_TYPES = 'This value should match at least one of these types [{types}].';
+
+type AllowedTypes = string | number | boolean | bigint | symbol | 'function' | object | 'array' | null | undefined;
```
Comment (imdhemy) [inline src/validator/constraints/basic/type.ts]: The allowed types should be strings.
```diff
- string | number | boolean ...
+ 'string' | 'number' | 'boolean' ...
```

Also undefined is not an allowed type. undefined should be bypassed.

File: src/validator/constraints/basic/type.ts (hunk 3 above)
Comment (imdhemy) [reply src/validator/constraints/basic/type.ts]: A single test case with a clear intention is sufficient. For example, “it should bypass undefined value.”

You’ve added two test cases with different descriptions of behavior. 

https://github.com/EmanFateen/framework/blob/d120ee5c1b643da1101e7b399e83b0ab1a726837/src/validator/constraints/basic/type.test.ts#L64

https://github.com/EmanFateen/framework/blob/d120ee5c1b643da1101e7b399e83b0ab1a726837/src/validator/constraints/basic/type.test.ts#L115

**Review by imdhemy — CHANGES_REQUESTED** [review]

File: src/validator/constraints/basic/type.ts (hunk 3 above)
Comment (EmanFateen) [reply src/validator/constraints/basic/type.ts]: ~~I kept `returns no violations for undefined value` and removed the redundant one. 
I see `returns no violations for undefined value` more clear than “it should bypass undefined value.”~~

Done

File: src/validator/constraints/basic/type.ts (hunk 7 above)
Comment (EmanFateen) [reply src/validator/constraints/basic/type.ts]: ~~The options used in the
`const normalizedTypes: AllowedTypes[] = Array.isArray(options.type) ? options.type : [options.type];`

`options.message ?? getDefaultMessageWith(normalizedTypes)`
That's why I needed it

However I will inline it.~~

I don't need it now. I inlined it (Y)

**Review by imdhemy — APPROVED** [review]
@EmanFateen 
Thank you!

I made `ConstraintOptions` generic so generic schema code can stay flexible, while specific constraints can keep precise option shapes. The old index signature leaked into every constraint-specific options type; now only the default generic form is loose, and `ConstraintOptions<{ ... }>` keeps custom constraint options strict while still sharing `groups`.
